File: invoice_parser.py
"""
Invoice parser for n8n Code node.

This module turns raw OCR'd invoice text into structured JSON. It is written
without external dependencies so it can be copied directly into an n8n Python
Code node.
"""

import json
import re
from typing import Any, Dict, List, Optional


HEADER_SKIP = {
    "CUSTOMER ORDER",
    "NUMBER",
    "DEL ADV",
    "DEL",
    "PRICE UNIT DISCOUNT TOTAL VALUE V",
    "C",
    "DESCRIPTION QTY",
    "CODE RATE% GOODS TAX",
    "TOTAL VALUE",
}


def _clean_value(value: str) -> str:
    """Normalize whitespace and strip trailing punctuation."""
    return re.sub(r"\s+", " ", value).strip(" \n:")


def _parse_money(value: str) -> Optional[float]:
    value = value.replace(",", "").strip()
    try:
        return float(value)
    except ValueError:
        return None


def _parse_header(lines: List[str]) -> Dict[str, Any]:
    result: Dict[str, Any] = {}
    for line in lines:
        if line.startswith("Customer Account No:"):
            result["customer_account"] = _clean_value(line.split(":", 1)[1])
        if line.startswith("TAXPOINT/DATE:"):
            parts = line.split(":", 1)[1].strip().split()
            if parts:
                result["taxpoint_date"] = parts[0]
            match = re.search(r"Doc\. Count:\s*(\d+)\s*of\s*(\d+)", line)
            if match:
                result["doc_count"] = {
                    "index": int(match.group(1)),
                    "total": int(match.group(2)),
                }
        if line.startswith("INVOICE NO:"):
            result["invoice_no"] = _clean_value(line.split(":", 1)[1])
    return result


def _extract_block(lines: List[str], start_key: str, end_key: str) -> List[str]:
    block: List[str] = []
    recording = False
    for line in lines:
        if line == start_key:
            recording = True
            continue
        if recording:
            if line == end_key:
                break
            block.append(line)
    return block


def _find_section_index(lines: List[str], marker: str) -> int:
    for idx, line in enumerate(lines):
        if line == marker:
            return idx
    return -1


def _parse_addresses(lines: List[str]) -> Dict[str, List[str]]:
    invoice_address = _extract_block(lines, "INVOICE ADDRESS", "DELIVERY ADDRESS")
    delivery_start = _find_section_index(lines, "DELIVERY ADDRESS")
    delivery_address: List[str] = []
    if delivery_start != -1:
        for line in lines[delivery_start + 1 :]:
            if line.startswith("INVOICE NO:"):
                break
            delivery_address.append(line)
    return {
        "invoice_address": [_clean_value(l) for l in invoice_address if l.strip()],
        "delivery_address": [_clean_value(l) for l in delivery_address if l.strip()],
    }


def _is_header_line(line: str) -> bool:
    normalized = _clean_value(line).upper()
    return normalized in HEADER_SKIP


def _parse_item_values(line: str) -> Optional[Dict[str, Any]]:
    pattern = r"^(\d+)\s+([0-9.]+)\s+([A-Z]+)\s+([0-9.]+)\s+([0-9.]+)\s+(\d+)"
    match = re.match(pattern, line.strip())
    if not match:
        return None
    return {
        "quantity": int(match.group(1)),
        "unit_price": float(match.group(2)),
        "unit": match.group(3),
        "discount": float(match.group(4)),
        "line_total": float(match.group(5)),
        "tax_code": match.group(6),
    }


def _parse_items(lines: List[str]) -> List[Dict[str, Any]]:
    items: List[Dict[str, Any]] = []
    desc_buffer: List[str] = []
    part_no: Optional[str] = None
    i = 0
    items_started = False
    while i < len(lines):
        line = lines[i].strip()
        if not line:
            i += 1
            continue
        if line.startswith("SUBTOTAL") or line.startswith("CODE RATE%"):
            break
        if _is_header_line(line):
            if "DESCRIPTION QTY" in line.upper():
                items_started = True
                desc_buffer.clear()
            i += 1
            continue
        if not items_started:
            i += 1
            continue
        if line.startswith("Our part no."):
            part_no = _clean_value(line.split("Our part no.", 1)[1])
            # find quantity line
            qty_line = ""
            j = i + 1
            while j < len(lines):
                qty_line = lines[j].strip()
                j += 1
                if qty_line:
                    break
            parsed_values = _parse_item_values(qty_line)
            description = _clean_value(" ".join(desc_buffer))
            desc_buffer.clear()
            if parsed_values:
                item = {
                    "description": description,
                    "part_number": part_no,
                    **parsed_values,
                }
                items.append(item)
            i = j
            continue
        desc_buffer.append(line)
        i += 1
    return items


def _parse_subtotal(lines: List[str]) -> Optional[float]:
    for line in lines:
        if line.startswith("SUBTOTAL"):
            parts = line.split()
            if len(parts) >= 2:
                return _parse_money(parts[-1])
    return None


def _parse_summary(lines: List[str]) -> Optional[Dict[str, Any]]:
    code_line_idx = _find_section_index(lines, "CODE RATE% GOODS TAX")
    if code_line_idx == -1:
        return None
    if code_line_idx + 1 >= len(lines):
        return None
    values_line = lines[code_line_idx + 1].strip()
    parts = values_line.split()
    if len(parts) < 4:
        return None
    summary = {
        "code": parts[0],
        "rate_percent": _parse_money(parts[1]),
        "goods": _parse_money(parts[2]),
        "tax": _parse_money(parts[3]),
    }
    total_line = None
    for line in lines[code_line_idx + 2 :]:
        if line.startswith("TOTAL TAX:"):
            total_line = line
            break
    if total_line:
        total_parts = total_line.replace("TOTAL TAX:", "").split()
        if len(total_parts) >= 3:
            summary["total_ex_vat"] = _parse_money(total_parts[0])
            summary["total_tax"] = _parse_money(total_parts[1])
            summary["total_inc_vat"] = _parse_money(total_parts[2])
    return summary


def parse_invoice_text(raw_text: str) -> List[Dict[str, Any]]:
    """
    Parse the supplied invoice text into a list of document dictionaries.
    This function is dependency free and safe to paste into an n8n Code node.
    """
    documents: List[Dict[str, Any]] = []
    blocks = [block.strip() for block in raw_text.split("SALES INVOICE") if block.strip()]
    for block in blocks:
        lines = [_clean_value(line) for line in block.splitlines() if line.strip()]
        header = _parse_header(lines)
        addresses = _parse_addresses(lines)
        items = _parse_items(lines)
        subtotal = _parse_subtotal(lines)
        summary = _parse_summary(lines)
        document: Dict[str, Any] = {
            **header,
            **addresses,
            "items": items,
        }
        if subtotal is not None:
            document["subtotal"] = subtotal
        if summary:
            document["summary"] = summary
        documents.append(document)
    return documents


def parse_to_json(raw_text: str) -> str:
    """Helper for n8n: returns a JSON string."""
    return json.dumps(parse_invoice_text(raw_text), indent=2)


def parse_to_items(raw_text: str) -> List[Dict[str, Any]]:
    """
    Helper for n8n: wraps each parsed document in the shape `{"json": ...}`
    that Code nodes expect when returning multiple items.
    """

    return [{"json": doc} for doc in parse_invoice_text(raw_text)]


if __name__ == "__main__":
    import sys

    if sys.stdin.isatty():
        print("Paste or pipe invoice text into stdin.")
    input_text = sys.stdin.read()
    print(parse_to_json(input_text))

File: invoice_parser.js
"use strict";

/**
 * Invoice parser for n8n Code node (JavaScript).
 *
 * Paste this file into an n8n Code node to convert raw OCR'd invoice text into
 * structured JSON. It is dependency free and mirrors the Python parser.
 */

const HEADER_SKIP = new Set([
  "CUSTOMER ORDER",
  "NUMBER",
  "DEL ADV",
  "DEL",
  "PRICE UNIT DISCOUNT TOTAL VALUE V",
  "C",
  "DESCRIPTION QTY",
  "CODE RATE% GOODS TAX",
  "TOTAL VALUE",
]);

function cleanValue(value) {
  return value.replace(/\s+/g, " ").replace(/[ \n:]+$/g, "").trim();
}

function parseMoney(value) {
  const normalized = value.replace(/,/g, "").trim();
  const num = Number.parseFloat(normalized);
  return Number.isNaN(num) ? null : num;
}

function parseHeader(lines) {
  const result = {};
  for (const line of lines) {
    if (line.startsWith("Customer Account No:")) {
      const idx = line.indexOf(":");
      if (idx !== -1) {
        result.customer_account = cleanValue(line.slice(idx + 1));
      }
    }
    if (line.startsWith("TAXPOINT/DATE:")) {
      const colonIdx = line.indexOf(":");
      const afterColon = colonIdx === -1 ? "" : line.slice(colonIdx + 1).trim();
      const parts = afterColon.split(/\s+/);
      if (parts.length > 0) {
        result.taxpoint_date = parts[0];
      }
      const match = line.match(/Doc\. Count:\s*(\d+)\s*of\s*(\d+)/);
      if (match) {
        result.doc_count = { index: Number(match[1]), total: Number(match[2]) };
      }
    }
    if (line.startsWith("INVOICE NO:")) {
      const idx = line.indexOf(":");
      if (idx !== -1) {
        result.invoice_no = cleanValue(line.slice(idx + 1));
      }
    }
  }
  return result;
}

function extractBlock(lines, startKey, endKey) {
  const block = [];
  let recording = false;
  for (const line of lines) {
    if (line === startKey) {
      recording = true;
      continue;
    }
    if (recording) {
      if (line === endKey) break;
      block.push(line);
    }
  }
  return block;
}

function findSectionIndex(lines, marker) {
  return lines.findIndex((line) => line === marker);
}

function parseAddresses(lines) {
  const invoiceAddress = extractBlock(lines, "INVOICE ADDRESS", "DELIVERY ADDRESS");
  const deliveryStart = findSectionIndex(lines, "DELIVERY ADDRESS");
  const deliveryAddress = [];
  if (deliveryStart !== -1) {
    for (let i = deliveryStart + 1; i < lines.length; i += 1) {
      const line = lines[i];
      if (line.startsWith("INVOICE NO:")) break;
      deliveryAddress.push(line);
    }
  }
  return {
    invoice_address: invoiceAddress.map((l) => cleanValue(l)).filter((l) => l.trim()),
    delivery_address: deliveryAddress.map((l) => cleanValue(l)).filter((l) => l.trim()),
  };
}

function isHeaderLine(line) {
  const normalized = cleanValue(line).toUpperCase();
  return HEADER_SKIP.has(normalized);
}

function parseItemValues(line) {
  const pattern = /^(\d+)\s+([0-9.]+)\s+([A-Z]+)\s+([0-9.]+)\s+([0-9.]+)\s+(\d+)/;
  const match = line.trim().match(pattern);
  if (!match) return null;
  return {
    quantity: Number(match[1]),
    unit_price: Number(match[2]),
    unit: match[3],
    discount: Number(match[4]),
    line_total: Number(match[5]),
    tax_code: Number(match[6]),
  };
}

function parseItems(lines) {
  const items = [];
  const descBuffer = [];
  let partNo = null;
  let i = 0;
  let itemsStarted = false;
  while (i < lines.length) {
    const line = lines[i].trim();
    if (!line) {
      i += 1;
      continue;
    }
    if (line.startsWith("SUBTOTAL") || line.startsWith("CODE RATE%")) {
      break;
    }
    if (isHeaderLine(line)) {
      if (line.toUpperCase().includes("DESCRIPTION QTY")) {
        itemsStarted = true;
        descBuffer.length = 0;
      }
      i += 1;
      continue;
    }
    if (!itemsStarted) {
      i += 1;
      continue;
    }
    if (line.startsWith("Our part no.")) {
      partNo = cleanValue(line.split("Our part no.").pop() || "");
      let qtyLine = "";
      let j = i + 1;
      while (j < lines.length) {
        qtyLine = lines[j].trim();
        j += 1;
        if (qtyLine) break;
      }
      const parsedValues = parseItemValues(qtyLine);
      const description = cleanValue(descBuffer.join(" "));
      descBuffer.length = 0;
      if (parsedValues) {
        items.push({ description, part_number: partNo, ...parsedValues });
      }
      i = j;
      continue;
    }
    descBuffer.push(line);
    i += 1;
  }
  return items;
}

function parseSubtotal(lines) {
  for (const line of lines) {
    if (line.startsWith("SUBTOTAL")) {
      const parts = line.split(/\s+/);
      if (parts.length >= 2) return parseMoney(parts[parts.length - 1]);
    }
  }
  return null;
}

function parseSummary(lines) {
  const codeIdx = findSectionIndex(lines, "CODE RATE% GOODS TAX");
  if (codeIdx === -1 || codeIdx + 1 >= lines.length) return null;
  const valuesLine = lines[codeIdx + 1].trim();
  const parts = valuesLine.split(/\s+/);
  if (parts.length < 4) return null;
  const summary = {
    code: parts[0],
    rate_percent: parseMoney(parts[1]),
    goods: parseMoney(parts[2]),
    tax: parseMoney(parts[3]),
  };
  const totalLine = lines.slice(codeIdx + 2).find((line) => line.startsWith("TOTAL TAX:"));
  if (totalLine) {
    const totalParts = totalLine.replace("TOTAL TAX:", "").trim().split(/\s+/);
    if (totalParts.length >= 3) {
      summary.total_ex_vat = parseMoney(totalParts[0]);
      summary.total_tax = parseMoney(totalParts[1]);
      summary.total_inc_vat = parseMoney(totalParts[2]);
    }
  }
  return summary;
}

function parseInvoiceText(rawText) {
  const documents = [];
  const blocks = rawText
    .split("SALES INVOICE")
    .map((block) => block.trim())
    .filter((block) => block);
  for (const block of blocks) {
    const lines = block
      .split(/\r?\n/)
      .map((line) => cleanValue(line))
      .filter((line) => line.trim());
    const header = parseHeader(lines);
    const addresses = parseAddresses(lines);
    const items = parseItems(lines);
    const subtotal = parseSubtotal(lines);
    const summary = parseSummary(lines);
    const document = {
      ...header,
      ...addresses,
      items,
    };
    if (subtotal !== null) document.subtotal = subtotal;
    if (summary) document.summary = summary;
    documents.push(document);
  }
  return documents;
}

function parseToJson(rawText) {
  return JSON.stringify(parseInvoiceText(rawText), null, 2);
}

function parseToItems(rawText) {
  return parseInvoiceText(rawText).map((doc) => ({ json: doc }));
}

if (require.main === module) {
  const fs = require("fs");
  const input = fs.readFileSync(0, "utf8");
  process.stdout.write(parseToJson(input));
}

module.exports = { parseInvoiceText, parseToJson, parseToItems };
